File: ios/bitchat/Views/Sonar/SonarOnboardingScreen.swift
//
// SonarOnboardingScreen.swift
// bitchat
//
// 3-step onboarding, ported from the Onboarding component in
// design/handoff/project/sonar/screens.jsx.
//
// This is free and unencumbered software released into the public domain.
// For more information, see <https://unlicense.org>
//

import SwiftUI
#if canImport(UIKit)
import UIKit
#elseif canImport(AppKit)
import AppKit
#endif

struct SonarOnboardingScreen: View {
    @EnvironmentObject private var store: SonarAppStore

    @State private var step = 0
    @State private var nick = ""
    @FocusState private var nickFocused: Bool
    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    // "I already have a key" — restore an existing identity from an nsec backup.
    @State private var restoring = false
    @State private var nsec = ""
    @State private var restoreError: String?
    @State private var restoreInFlight = false
    @FocusState private var nsecFocused: Bool

    private var can: Bool { nick.trimmingCharacters(in: .whitespaces).count >= 2 }
    private var trimmedNick: String { nick.trimmingCharacters(in: .whitespaces) }

    /// Matches the prototype's nsec validation: bech32 `nsec1…`.
    private var nsecOk: Bool {
        let t = nsec.trimmingCharacters(in: .whitespacesAndNewlines)
        return t.range(of: "^nsec1[0-9a-z]{20,}$", options: .regularExpression) != nil
    }

    /// Nickname suggestions from the design handoff (data.js) — design copy,
    /// not demo data: tapping "Surprise me" picks one as the user's real nick.
    private static let suggestions = ["quietfox", "tram12", "lakeswim", "verdigris", "morningstatic", "papercrane", "northpine", "softsignal"]

    private func surprise() {
        nick = Self.suggestions.randomElement() ?? "quietfox"
    }

    var body: some View {
        VStack(spacing: 0) {
            // bc-obtop
            HStack {
                if step > 0 || restoring {
                    SNIconButton(action: {
                        if restoring { restoring = false; restoreError = nil }
                        else { advance(to: step - 1) }
                    }) {
                        SNIcon(name: .back, size: 21, weight: 2.1)
                    }
                }
                Spacer()
            }
            .frame(height: 40)
            .padding(.leading, -10)

            // bc-obbody
            Group {
                if restoring {
                    stepRestore
                } else {
                    switch step {
                    case 0: stepIntro
                    case 1: stepNickname
                    default: stepDone
                    }
                }
            }
            .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .leading)
            .transition(reduceMotion ? .opacity : .offset(y: 10).combined(with: .opacity))
            .id(restoring ? -1 : step)

            // bc-obfooter
            VStack(spacing: 16) {
                if !restoring {
                    HStack(spacing: 6) {
                        ForEach(0..<3, id: \.self) { i in
                            Circle()
                                .fill(step == i ? SonarTheme.accent : SonarTheme.hairline)
                                .frame(width: 7, height: 7)
                        }
                    }
                }
                if restoring {
                    SNPrimaryButton(label: restoreInFlight ? "Restoring\u{2026}" : "Restore account", disabled: !nsecOk || restoreInFlight) {
                        restore()
                    }
                } else {
                    switch step {
                    case 0:
                        SNPrimaryButton(label: "Get started") { advance(to: 1) }
                        // Full-width accent soft CTA — ghost text was too easy to miss
                        // for returning users restoring with an nsec.
                        Button(action: {
                            nsec = ""
                            restoreError = nil
                            withAnimation(reduceMotion ? nil : .easeOut(duration: 0.35)) { restoring = true }
                        }) {
                            Text("Restore account with private key")
                                .font(SonarTheme.uiFont(size: 16, weight: .bold))
                                .foregroundColor(SonarTheme.accentDeep)
                                .frame(maxWidth: .infinity)
                                .padding(.vertical, 14)
                                .background(
                                    RoundedRectangle(cornerRadius: 16, style: .continuous)
                                        .fill(SonarTheme.accentSoft)
                                )
                        }
                        .buttonStyle(SNScaleStyle(scale: 0.98))
                        .padding(.top, 4)
                    case 1:
                        SNPrimaryButton(label: "Continue", disabled: !can) { advance(to: 2) }
                    default:
                        SNPrimaryButton(label: "Start chatting") {
                            store.completeOnboarding(nick: trimmedNick)
                        }
                    }
                }
            }
            .padding(.top, 18)
        }
        .padding(EdgeInsets(top: 10, leading: 28, bottom: 12, trailing: 28))
        .frame(maxWidth: .infinity, maxHeight: .infinity)
        .background(SonarTheme.bg.ignoresSafeArea())
    }

    private func advance(to newStep: Int) {
        if reduceMotion {
            step = newStep
        } else {
            withAnimation(.easeOut(duration: 0.35)) { step = newStep }
        }
    }

    // Step 1 — intro
    private var stepIntro: some View {
        VStack(alignment: .leading, spacing: 0) {
            Spacer()
            RoundedRectangle(cornerRadius: 23, style: .continuous)
                .fill(SonarTheme.accentFill)
                .frame(width: 74, height: 74)
                .overlay(
                    SNIcon(name: .rings, size: 40, weight: 1.5)
                        .foregroundColor(SonarTheme.onAccent)
                )
                .padding(.bottom, 28)
            Text("Sense who\u{2019}s nearby before you see them.")
                .font(SonarTheme.uiFont(size: 30, weight: .heavy))
                .kerning(-30 * 0.02)
                .foregroundColor(SonarTheme.text)
                .padding(.bottom, 10)
            Text("Sonar connects phones directly — no phone number, no account, no servers.")
                .font(SonarTheme.uiFont(size: 16))
                .lineSpacing(16 * 0.3)
                .foregroundColor(SonarTheme.text2)
                .padding(.bottom, 22)
            featureRow(icon: .mesh, title: "Works without internet", desc: "Bluetooth finds people around you, even offline.")
            featureRow(icon: .globe, title: "Out of range? Still reachable", desc: "Messages travel encrypted over the open internet instead.")
            featureRow(icon: .lock, title: "Private by design", desc: "Direct messages are end-to-end encrypted. Always.")
            Spacer()
        }
    }

    private func featureRow(icon: SNIconName, title: String, desc: String) -> some View {
        HStack(alignment: .top, spacing: 14) {
            RoundedRectangle(cornerRadius: 13, style: .continuous)
                .fill(SonarTheme.accentSoft)
                .frame(width: 40, height: 40)
                .overlay(
                    SNIcon(name: icon, size: 20)
                        .foregroundColor(SonarTheme.accentDeep)
                )
            VStack(alignment: .leading, spacing: 2) {
                Text(verbatim: title)
                    .font(SonarTheme.uiFont(size: 16, weight: .bold))
                    .foregroundColor(SonarTheme.text)
                Text(verbatim: desc)
                    .font(SonarTheme.uiFont(size: 13.5))
                    .lineSpacing(13.5 * 0.25)
                    .foregroundColor(SonarTheme.text2)
            }
        }
        .padding(.vertical, 11)
    }

    // Step 2 — nickname
    private var stepNickname: some View {
        VStack(alignment: .leading, spacing: 0) {
            Spacer()
            Text("Pick a nickname")
                .font(SonarTheme.uiFont(size: 30, weight: .heavy))
                .kerning(-30 * 0.02)
                .foregroundColor(SonarTheme.text)
                .padding(.bottom, 10)
            Text("It\u{2019}s just what people see — change it anytime.")
                .font(SonarTheme.uiFont(size: 16))
                .lineSpacing(16 * 0.3)
                .foregroundColor(SonarTheme.text2)
                .padding(.bottom, 22)
            HStack(spacing: 16) {
                SonarAvatar(name: trimmedNick.isEmpty ? "?" : trimmedNick, size: 72)
                TextField("", text: $nick, prompt: Text("nickname").foregroundColor(SonarTheme.text3).fontWeight(.medium))
                    .textFieldStyle(.plain)
                    .font(SonarTheme.uiFont(size: 21, weight: .bold))
                    .foregroundColor(SonarTheme.text)
                    .focused($nickFocused)
                    .onSubmit { if can { advance(to: 2) } }
                    .onChange(of: nick) { v in
                        if v.count > 20 { nick = String(v.prefix(20)) }
                    }
                    .padding(EdgeInsets(top: 15, leading: 16, bottom: 15, trailing: 16))
                    .background(RoundedRectangle(cornerRadius: 16, style: .continuous).fill(SonarTheme.surface2))
                    .overlay(
                        RoundedRectangle(cornerRadius: 16, style: .continuous)
                            .strokeBorder(nickFocused ? SonarTheme.accent : Color.clear, lineWidth: 1.5)
                    )
            }
            .padding(.bottom, 18)
            Button(action: surprise) {
                HStack(spacing: 7) {
                    SNIcon(name: .dice, size: 16, weight: 2)
                    Text("Surprise me")
                        .font(SonarTheme.uiFont(size: 14, weight: .bold))
                }
                .foregroundColor(SonarTheme.accentDeep)
                .padding(EdgeInsets(top: 9, leading: 14, bottom: 9, trailing: 14))
                .background(Capsule().fill(SonarTheme.accentSoft))
            }
            .buttonStyle(SNScaleStyle(scale: 0.96))
            .padding(.top, 12)
            Text("No signup. Your identity is a private key created on this phone — nobody else ever sees it.")
                .font(SonarTheme.uiFont(size: 13))
                .lineSpacing(13 * 0.3)
                .foregroundColor(SonarTheme.text3)
                .padding(.top, 18)
            Spacer()
        }
    }

    // Step 3 — you're in
    private var stepDone: some View {
        VStack(alignment: .leading, spacing: 0) {
            Spacer()
            SonarAvatar(name: trimmedNick, size: 92)
                .padding(.bottom, 22)
            Text(verbatim: "You\u{2019}re in, \(trimmedNick).")
                .font(SonarTheme.uiFont(size: 30, weight: .heavy))
                .kerning(-30 * 0.02)
                .foregroundColor(SonarTheme.text)
                .padding(.bottom, 10)
            Text("No account was created anywhere — your identity lives on this phone.")
                .font(SonarTheme.uiFont(size: 16))
                .lineSpacing(16 * 0.3)
                .foregroundColor(SonarTheme.text2)
            SNFingerprintCard(label: "Your key fingerprint", value: store.myFingerprintDisplay)
                .padding(.top, 24)
            Text("Friends can verify this fingerprint in person to be sure it\u{2019}s really you.")
                .font(SonarTheme.uiFont(size: 13))
                .lineSpacing(13 * 0.3)
                .foregroundColor(SonarTheme.text3)
                .padding(.top, 18)
            Spacer()
        }
    }

    // Restore — paste an existing nsec ("I already have a key")
    private var stepRestore: some View {
        VStack(alignment: .leading, spacing: 0) {
            Spacer()
            RoundedRectangle(cornerRadius: 23, style: .continuous)
                .fill(SonarTheme.accentFill)
                .frame(width: 74, height: 74)
                .overlay(
                    SNIcon(name: .importKey, size: 34, weight: 1.7)
                        .foregroundColor(SonarTheme.onAccent)
                )
                .padding(.bottom, 28)
            Text("Restore your account")
                .font(SonarTheme.uiFont(size: 30, weight: .heavy))
                .kerning(-30 * 0.02)
                .foregroundColor(SonarTheme.text)
                .padding(.bottom, 10)
            (Text("Paste your ") + Text("nsec").fontWeight(.bold)
                + Text(" private key. This restores your Sonar identity and Lightning wallet on this phone. Chat history on this device starts fresh until backup ships."))
                .font(SonarTheme.uiFont(size: 16))
                .lineSpacing(16 * 0.3)
                .foregroundColor(SonarTheme.text2)
                .fixedSize(horizontal: false, vertical: true)
                .padding(.bottom, 22)
            ZStack(alignment: .topLeading) {
                if nsec.isEmpty {
                    Text(verbatim: "nsec1\u{2026}")
                        .font(SonarTheme.monoFont(size: 15))
                        .foregroundColor(SonarTheme.text3)
                        .padding(EdgeInsets(top: 14, leading: 16, bottom: 0, trailing: 16))
                        .allowsHitTesting(false)
                }
                TextEditor(text: $nsec)
                    .font(SonarTheme.monoFont(size: 15))
                    .foregroundColor(SonarTheme.text)
                    .focused($nsecFocused)
                    .frame(minHeight: 84)
                    .scrollContentBackground(.hidden)
                    .padding(EdgeInsets(top: 6, leading: 11, bottom: 6, trailing: 11))
                    #if os(iOS)
                    .textInputAutocapitalization(.never)
                    .autocorrectionDisabled()
                    #endif
            }
            .background(RoundedRectangle(cornerRadius: 16, style: .continuous).fill(SonarTheme.surface2))
            .overlay(
                RoundedRectangle(cornerRadius: 16, style: .continuous)
                    .strokeBorder(nsecFocused ? SonarTheme.accent : Color.clear, lineWidth: 1.5)
            )
            Button(action: pasteFromClipboard) {
                HStack(spacing: 7) {
                    SNIcon(name: .copy, size: 15, weight: 2)
                    Text("Paste from clipboard")
                        .font(SonarTheme.uiFont(size: 14, weight: .bold))
                }
                .foregroundColor(SonarTheme.accentDeep)
                .padding(EdgeInsets(top: 9, leading: 14, bottom: 9, trailing: 14))
                .background(Capsule().fill(SonarTheme.accentSoft))
            }
            .buttonStyle(SNScaleStyle(scale: 0.96))
            .padding(.top, 12)
            if let restoreError {
                Text(verbatim: restoreError)
                    .font(SonarTheme.uiFont(size: 13))
                    .lineSpacing(13 * 0.3)
                    .foregroundColor(SonarTheme.danger)
                    .padding(.top, 12)
            }
            Text("Sonar never sends this key anywhere — it\u{2019}s decoded on this phone to unlock your identity.")
                .font(SonarTheme.uiFont(size: 13))
                .lineSpacing(13 * 0.3)
                .foregroundColor(SonarTheme.text3)
                .padding(.top, 18)
            Spacer()
        }
    }

    private func pasteFromClipboard() {
        #if canImport(UIKit)
        if let s = UIPasteboard.general.string {
            nsec = s.trimmingCharacters(in: .whitespacesAndNewlines)
        }
        #elseif canImport(AppKit)
        if let s = NSPasteboard.general.string(forType: .string) {
            nsec = s.trimmingCharacters(in: .whitespacesAndNewlines)
        }
        #endif
    }

    private func restore() {
        let key = nsec.trimmingCharacters(in: .whitespacesAndNewlines)
        guard nsecOk, !restoreInFlight else { return }
        restoreInFlight = true
        restoreError = nil
        Task { @MainActor in
            do {
                try await store.restoreAccount(nsec: key)
                // store.restoreAccount flips `onboarded`; the root view swaps to home.
            } catch {
                restoreError = (error as? SonarAccountRestoreError)?.localizedDescription
                    ?? "That key couldn\u{2019}t be imported. Check you pasted the full nsec1\u{2026} key."
                restoreInFlight = false
            }
        }
    }
}

/// bc-fpcard — fingerprint card used by onboarding step 3 and the profile screen.
struct SNFingerprintCard: View {
    let label: String
    let value: String

    var body: some View {
        VStack(alignment: .leading, spacing: 4) {
            Text(verbatim: label.uppercased())
                .font(SonarTheme.uiFont(size: 12, weight: .bold))
                .kerning(12 * 0.05)
                .foregroundColor(SonarTheme.text3)
            Text(verbatim: value)
                .font(SonarTheme.monoFont(size: 15))
                .kerning(15 * 0.08)
                .foregroundColor(SonarTheme.text)
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(EdgeInsets(top: 13, leading: 16, bottom: 13, trailing: 16))
        .background(RoundedRectangle(cornerRadius: 14, style: .continuous).fill(SonarTheme.surface2))
    }
}

File: apps/sonar/composeApp/src/commonMain/kotlin/chat/bitchat/sonar/screens/SonarOnboardingScreen.kt
package chat.bitchat.sonar.screens

import androidx.compose.animation.AnimatedContent
import androidx.compose.animation.fadeIn
import androidx.compose.animation.fadeOut
import androidx.compose.animation.togetherWith
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.LocalTextStyle
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.platform.LocalClipboardManager
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import chat.bitchat.sonar.SonarAccountRestoreException
import chat.bitchat.sonar.SonarAppState
import chat.bitchat.sonar.resources.Res
import chat.bitchat.sonar.resources.anyone_with_this_key_controls_your
import chat.bitchat.sonar.resources.bluetooth_finds_people_around_you_even
import chat.bitchat.sonar.resources.continue_
import chat.bitchat.sonar.resources.create_a_new_identity
import chat.bitchat.sonar.resources.direct_messages_are_end_to_end
import chat.bitchat.sonar.resources.friends_can_verify_this_fingerprint_in
import chat.bitchat.sonar.resources.generating
import chat.bitchat.sonar.resources.get_started
import chat.bitchat.sonar.resources.it_s_just_what_people_see_change_it
import chat.bitchat.sonar.resources.messages_travel_encrypted_over_the_open
import chat.bitchat.sonar.resources.nickname
import chat.bitchat.sonar.resources.no_account_was_created_anywhere_your
import chat.bitchat.sonar.resources.no_signup_your_identity_is_a_private
import chat.bitchat.sonar.resources.nsec1
import chat.bitchat.sonar.resources.out_of_range_still_reachable
import chat.bitchat.sonar.resources.paste_key
import chat.bitchat.sonar.resources.paste_your_nsec_private_key_to_bring
import chat.bitchat.sonar.resources.pick_a_nickname
import chat.bitchat.sonar.resources.private_by_design
import chat.bitchat.sonar.resources.restore_account
import chat.bitchat.sonar.resources.restoring
import chat.bitchat.sonar.resources.sense_who_s_nearby_before_you_see_them
import chat.bitchat.sonar.resources.sonar_connects_phones_directly_no_phone
import chat.bitchat.sonar.resources.start_chatting
import chat.bitchat.sonar.resources.surprise_me
import chat.bitchat.sonar.resources.that_key_couldn_t_be_imported_check_you
import chat.bitchat.sonar.resources.works_without_internet
import chat.bitchat.sonar.resources.you_re_in
import chat.bitchat.sonar.resources.your_key_fingerprint_2
import chat.bitchat.sonar.ui.SNFingerprintCard
import chat.bitchat.sonar.ui.SNGhostButton
import chat.bitchat.sonar.ui.SNIcon
import chat.bitchat.sonar.ui.SNIconButton
import chat.bitchat.sonar.ui.SNIconName
import chat.bitchat.sonar.ui.SNPrimaryButton
import chat.bitchat.sonar.ui.SonarAvatar
import chat.bitchat.sonar.ui.sonar
import org.jetbrains.compose.resources.stringResource

private val SUGGESTIONS = listOf(
    "quietfox", "tram12", "lakeswim", "verdigris", "morningstatic", "papercrane", "northpine", "softsignal"
)

@Composable
fun SonarOnboardingScreen(state: SonarAppState) {
    val s = sonar
    var step by remember { mutableStateOf(0) }
    var nick by remember { mutableStateOf("") }
    var restoring by remember { mutableStateOf(false) }
    var nsec by remember { mutableStateOf("") }
    var restoreError by remember { mutableStateOf<String?>(null) }
    var restoreInFlight by remember { mutableStateOf(false) }
    val trimmed = nick.trim()
    val can = trimmed.length >= 2
    val nsecOk = nsec.trim().matches(Regex("^nsec1[0-9a-z]{20,}$"))
    val clipboard = LocalClipboardManager.current
    val restoreFailureFallback =
        stringResource(Res.string.that_key_couldn_t_be_imported_check_you)

    Column(
        Modifier.fillMaxSize().background(s.bg)
            .padding(start = 28.dp, end = 28.dp, top = 10.dp, bottom = 12.dp)
    ) {
        // Top bar (back)
        Box(Modifier.fillMaxWidth().height(40.dp)) {
            if (step > 0 || restoring) {
                SNIconButton(SNIconName.Back, onClick = {
                    if (restoring) {
                        restoring = false
                        restoreError = null
                    } else {
                        step -= 1
                    }
                })
            }
        }

        AnimatedContent(
            targetState = if (restoring) -1 else step,
            transitionSpec = { fadeIn() togetherWith fadeOut() },
            modifier = Modifier.weight(1f)
        ) { st ->
            when (st) {
                -1 -> StepRestore(
                    nsec = nsec,
                    error = restoreError,
                    onChange = { nsec = it },
                    onPaste = { clipboard.getText()?.text?.let { nsec = it.trim() } },
                )
                0 -> StepIntro()
                1 -> StepNickname(nick, trimmed) { nick = it.take(20) }
                else -> StepDone(trimmed, state.fingerprint())
            }
        }

        // Footer
        Column(Modifier.padding(top = 18.dp), horizontalAlignment = Alignment.CenterHorizontally) {
            if (!restoring) {
                Row(horizontalArrangement = Arrangement.spacedBy(6.dp)) {
                    repeat(3) { i ->
                        Box(
                            Modifier.size(7.dp).clip(CircleShape)
                                .background(if (step == i) s.accent else s.hairline)
                        )
                    }
                }
                Spacer(Modifier.height(16.dp))
            }
            if (restoring) {
                SNPrimaryButton(
                    if (restoreInFlight) stringResource(Res.string.restoring) else stringResource(Res.string.restore_account),
                    disabled = !nsecOk || restoreInFlight,
                ) {
                    val key = nsec.trim()
                    restoreInFlight = true
                    restoreError = null
                    state.restoreAccount(key) { result ->
                        restoreInFlight = false
                        result.exceptionOrNull()?.let { failure ->
                            restoreError = (failure as? SonarAccountRestoreException)?.message
                                ?: restoreFailureFallback
                        }
                    }
                }
                SNGhostButton(stringResource(Res.string.create_a_new_identity)) {
                    restoring = false
                    restoreError = null
                    nsec = ""
                }
            } else {
                when (step) {
                    0 -> {
                        // Keep both first-run paths in one bounded row. A stacked
                        // secondary action was clipped below the viewport on the
                        // API 36 medium-phone layout, making restore undiscoverable.
                        Row(
                            Modifier.fillMaxWidth(),
                            horizontalArrangement = Arrangement.spacedBy(8.dp),
                        ) {
                            SNPrimaryButton(stringResource(Res.string.get_started), Modifier.weight(1f)) { step = 1 }
                            Box(
                                Modifier.weight(1f).height(52.dp)
                                    .clip(RoundedCornerShape(15.dp)).background(s.accentSoft)
                                    .clickable {
                                        nsec = ""
                                        restoreError = null
                                        restoring = true
                                    },
                                contentAlignment = Alignment.Center,
                            ) {
                                Text(
                                    stringResource(Res.string.restore_account),
                                    color = s.accentDeep,
                                    fontSize = 15.sp,
                                    fontWeight = FontWeight.Bold,
                                )
                            }
                        }
                    }
                    1 -> SNPrimaryButton(stringResource(Res.string.continue_), disabled = !can) { if (can) step = 2 }
                    else -> SNPrimaryButton(stringResource(Res.string.start_chatting)) { state.completeOnboarding(trimmed) }
                }
            }
        }
    }
}

@Composable
private fun StepIntro() {
    val s = sonar
    Column(Modifier.fillMaxSize(), verticalArrangement = Arrangement.Center) {
        Box(
            Modifier.size(74.dp).clip(RoundedCornerShape(23.dp)).background(s.accentFill),
            contentAlignment = Alignment.Center
        ) { SNIcon(SNIconName.Rings, 40.dp, s.onAccent, weight = 1.5f) }
        Spacer(Modifier.height(28.dp))
        Text(
            stringResource(Res.string.sense_who_s_nearby_before_you_see_them),
            color = s.text, fontSize = 30.sp, fontWeight = FontWeight.Black, lineHeight = 34.sp
        )
        Spacer(Modifier.height(10.dp))
        Text(
            stringResource(Res.string.sonar_connects_phones_directly_no_phone),
            color = s.text2, fontSize = 16.sp, lineHeight = 21.sp
        )
        Spacer(Modifier.height(22.dp))
        FeatureRow(SNIconName.Mesh, stringResource(Res.string.works_without_internet), stringResource(Res.string.bluetooth_finds_people_around_you_even))
        FeatureRow(SNIconName.Globe, stringResource(Res.string.out_of_range_still_reachable), stringResource(Res.string.messages_travel_encrypted_over_the_open))
        FeatureRow(SNIconName.Lock, stringResource(Res.string.private_by_design), stringResource(Res.string.direct_messages_are_end_to_end))
    }
}

@Composable
private fun FeatureRow(icon: SNIconName, title: String, desc: String) {
    val s = sonar
    Row(Modifier.fillMaxWidth().padding(vertical = 11.dp)) {
        Box(
            Modifier.size(40.dp).clip(RoundedCornerShape(13.dp)).background(s.accentSoft),
            contentAlignment = Alignment.Center
        ) { SNIcon(icon, 20.dp, s.accentDeep) }
        Spacer(Modifier.width(14.dp))
        Column {
            Text(title, color = s.text, fontSize = 16.sp, fontWeight = FontWeight.Bold)
            Text(desc, color = s.text2, fontSize = 13.5.sp, lineHeight = 17.sp)
        }
    }
}

@Composable
private fun StepNickname(nick: String, trimmed: String, onChange: (String) -> Unit) {
    val s = sonar
    Column(Modifier.fillMaxSize(), verticalArrangement = Arrangement.Center) {
        Text(stringResource(Res.string.pick_a_nickname), color = s.text, fontSize = 30.sp, fontWeight = FontWeight.Black)
        Spacer(Modifier.height(10.dp))
        Text(stringResource(Res.string.it_s_just_what_people_see_change_it), color = s.text2, fontSize = 16.sp, lineHeight = 21.sp)
        Spacer(Modifier.height(22.dp))
        Row(verticalAlignment = Alignment.CenterVertically) {
            SonarAvatar(trimmed.ifEmpty { "?" }, 72.dp)
            Spacer(Modifier.width(16.dp))
            Box(
                Modifier.weight(1f).clip(RoundedCornerShape(16.dp)).background(s.surface2)
                    .padding(horizontal = 16.dp, vertical = 15.dp)
            ) {
                if (nick.isEmpty()) Text(stringResource(Res.string.nickname), color = s.text3, fontSize = 21.sp, fontWeight = FontWeight.Medium)
                BasicTextField(
                    value = nick,
                    onValueChange = onChange,
                    singleLine = true,
                    textStyle = TextStyle(color = s.text, fontSize = 21.sp, fontWeight = FontWeight.Bold),
                    cursorBrush = SolidColor(s.accent),
                    modifier = Modifier.fillMaxWidth()
                )
            }
        }
        Spacer(Modifier.height(18.dp))
        Row(
            Modifier.clip(CircleShape).background(s.accentSoft).clickable {
                onChange(SUGGESTIONS[(nick.hashCode() and 0x7fffffff) % SUGGESTIONS.size])
            }.padding(horizontal = 14.dp, vertical = 9.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            SNIcon(SNIconName.Dice, 16.dp, s.accentDeep, weight = 2f)
            Spacer(Modifier.width(7.dp))
            Text(stringResource(Res.string.surprise_me), color = s.accentDeep, fontSize = 14.sp, fontWeight = FontWeight.Bold)
        }
        Spacer(Modifier.height(18.dp))
        Text(
            stringResource(Res.string.no_signup_your_identity_is_a_private),
            color = s.text3, fontSize = 13.sp, lineHeight = 17.sp
        )
    }
}

@Composable
private fun StepRestore(
    nsec: String,
    error: String?,
    onChange: (String) -> Unit,
    onPaste: () -> Unit,
) {
    val s = sonar
    Column(Modifier.fillMaxSize().verticalScroll(rememberScrollState()), verticalArrangement = Arrangement.Center) {
        Text(stringResource(Res.string.restore_account), color = s.text, fontSize = 30.sp, fontWeight = FontWeight.Black)
        Spacer(Modifier.height(10.dp))
        Text(
            stringResource(Res.string.paste_your_nsec_private_key_to_bring),
            color = s.text2, fontSize = 16.sp, lineHeight = 21.sp,
        )
        Spacer(Modifier.height(22.dp))
        Box(
            Modifier.fillMaxWidth().heightIn(min = 116.dp).clip(RoundedCornerShape(16.dp)).background(s.surface2)
                .padding(horizontal = 16.dp, vertical = 14.dp)
        ) {
            if (nsec.isEmpty()) Text(stringResource(Res.string.nsec1), color = s.text3, fontSize = 16.sp)
            BasicTextField(
                value = nsec,
                onValueChange = { onChange(it.trim()) },
                textStyle = TextStyle(color = s.text, fontSize = 15.sp, lineHeight = 20.sp),
                cursorBrush = SolidColor(s.accent),
                modifier = Modifier.fillMaxWidth()
            )
        }
        Spacer(Modifier.height(12.dp))
        Row(
            Modifier.clip(CircleShape).background(s.accentSoft).clickable(onClick = onPaste)
                .padding(horizontal = 14.dp, vertical = 9.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            SNIcon(SNIconName.Key, 16.dp, s.accentDeep, weight = 2f)
            Spacer(Modifier.width(7.dp))
            Text(stringResource(Res.string.paste_key), color = s.accentDeep, fontSize = 14.sp, fontWeight = FontWeight.Bold)
        }
        Spacer(Modifier.height(18.dp))
        Text(stringResource(Res.string.anyone_with_this_key_controls_your), color = s.text3, fontSize = 13.sp, lineHeight = 17.sp)
        if (error != null) {
            Spacer(Modifier.height(14.dp))
            Text(error, color = s.danger, fontSize = 13.sp, lineHeight = 17.sp)
        }
    }
}

@Composable
private fun StepDone(nick: String, fingerprint: String) {
    val s = sonar
    val generatingLabel = stringResource(Res.string.generating)
    Column(Modifier.fillMaxSize(), verticalArrangement = Arrangement.Center) {
        SonarAvatar(nick.ifEmpty { "?" }, 92.dp)
        Spacer(Modifier.height(22.dp))
        Text(stringResource(Res.string.you_re_in, nick), color = s.text, fontSize = 30.sp, fontWeight = FontWeight.Black)
        Spacer(Modifier.height(10.dp))
        Text(stringResource(Res.string.no_account_was_created_anywhere_your), color = s.text2, fontSize = 16.sp, lineHeight = 21.sp)
        Spacer(Modifier.height(24.dp))
        SNFingerprintCard(
            stringResource(Res.string.your_key_fingerprint_2),
            fingerprint.ifEmpty { generatingLabel },
        )
        Spacer(Modifier.height(18.dp))
        Text(stringResource(Res.string.friends_can_verify_this_fingerprint_in), color = s.text3, fontSize = 13.sp, lineHeight = 17.sp)
    }
}
